plugins/plugin-ssh-machine/src/main/java/org/eclipse/che/plugin/machine/ssh/SshMachineInstance.java

```diff
@@ -0,0 +1,199 @@
+import javax.ws.rs.core.UriBuilder;
+import java.net.URI;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+import static java.lang.String.format;
+import static java.util.Collections.emptyMap;
+import static java.util.stream.Collectors.toMap;
+
+/**
+ * Implementation of {@link Instance} that uses represents ssh machine.
+ *
+ * @author Alexander Garagatyi
+ *
+ * @see SshMachineInstanceProvider
+ */
+// todo try to avoid map of processes
+public class SshMachineInstance extends AbstractInstance {
+    private static final AtomicInteger pidSequence = new AtomicInteger(1);
+
+    private final SshClient                                   sshClient;
+    private final LineConsumer                                outputConsumer;
+    private final SshMachineFactory                           machineFactory;
+    private final Set<ServerConf>                             machinesServers;
+    private final ConcurrentHashMap<Integer, InstanceProcess> machineProcesses;
+
+    private MachineRuntimeInfoImpl machineRuntime;
+
+    @Inject
+    public SshMachineInstance(@Assisted Machine machine,
+                              @Assisted SshClient sshClient,
+                              @Assisted LineConsumer outputConsumer,
+                              SshMachineFactory machineFactory,
+                              @Named("machine.ssh.machine_servers") Set<ServerConf> machinesServers) {
+        super(machine);
+        this.sshClient = sshClient;
+        this.outputConsumer = outputConsumer;
+        this.machineFactory = machineFactory;
+        this.machinesServers = machinesServers;
+        this.machineProcesses = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public LineConsumer getLogger() {
+        return outputConsumer;
+    }
+
+    @Override
+    public MachineRuntimeInfoImpl getRuntime() {
+        // lazy initialization
+        if (machineRuntime == null) {
+            UriBuilder uriBuilder = UriBuilder.fromUri("http://" + sshClient.getHost());
+
+            Map<String, ServerImpl> servers = machinesServers.stream()
+                                                             .collect(toMap(ServerConf::getPort,
+                                                                            serverConf -> serverConfToServer(serverConf,
+                                                                                                             uriBuilder.clone())));
+            machineRuntime = new MachineRuntimeInfoImpl(emptyMap(), emptyMap(), servers);
+            // todo get env from client
+        }
+        return machineRuntime;
+    }
+
+    @Override
+    public InstanceProcess getProcess(final int pid) throws NotFoundException, MachineException {
+        final InstanceProcess machineProcess = machineProcesses.get(pid);
+        if (machineProcess != null) {
```

Comment:
You can avoid negation + extra indentation by throwing exception when `machineProcess == null`